### ISSUE TITLE
State of `master` versus `develop` branches

### ISSUE THREAD

**Issue by danielweck** [readium/readium-lcp-server#235]
Git commit history diff between `develop` ... `master`:
https://github.com/readium/readium-lcp-server/compare/develop...master
This should be empty, but it isn't. This indicates that `master` does not contain the Git commit graph history of `develop`, which is a sure sign that the branches are out of sync.

Git commit history diff between `master` ... `develop`:
https://github.com/readium/readium-lcp-server/compare/master...develop
Several changes here, which is fine as this indicates that `develop` is ahead of `master` with additional PR merges since the last official release.

Normally, `develop` contains the latest code (i.e. merged PRs, source tree not release-tagged yet), and `master` contains the exact commit graph history of `develop`, but stops at the latest release tag.

Example with Thorium:

Git commit history diff between `develop` ... `master` (empty):
https://github.com/edrlab/thorium-reader/compare/develop...master

Git commit history diff between `master` ... `develop` (lots of changes, due to the pending release):
https://github.com/edrlab/thorium-reader/compare/master...develop

**Comment by danielweck** [comment]
Possible remediation:

1) Ensure local clone is up to date with the `origin` remote, using `git pull` (or `git fetch` followed by `git merge`) in the `develop` and `master` branches, in isolation. `git status` / `git diff` should show a clean local copy.
2) `git checkout develop` followed by `git merge master`, to ensure that any future code changes (i.e. PR merged into `develop`) are a continuation of the `master` Git commit history.
3) When ready for an official release, `git checkout master` followed by `git merge develop` so that `master` now contains all the new stuff introduced in `develop`.
4) This is an opportunity to release-tag the HEAD commit in `master` (which by the way, should be exactly the same as in `develop`, therefore the release tag will also appear in the `develop` history), for example:
```
GIT_COMMITTER_DATE="$(git show master --format=%aD | head -1)" \
git tag -a "v1.0" master -m "tag v1.0" \
&& git push --tags origin master \
&& git tag --list --format='%(refname)   %(taggerdate)'
```

...in case of a mistake, the tag can be removed locally and on the server:

```
git tag -d v1.0 && git push origin :refs/tags/v1.0
```

**Comment by danielweck** [comment]
I would also add that BOTH `master` and `develop` (the former being the exact image of the latter, just stopping at an earlier point in history) should contain a  "clean" Git commit history, i.e. containing a sequence of atomic changesets labelled with a human-readable and machine-parseable commit message syntax. This is why it is important to merge PRs into the `develop` branch with a "squash", so that individual PR Git commits are not interspersed into the destination branch (typically with "merge commits" that mark discrete moments in history that "batch" correlated Git commits, but the resulting stream of commits from beginning to end remains really messy when not squashing at the point of PR merge.

**Comment by danielweck** [comment]
I would recommend using:
https://www.conventionalcommits.org/en/v1.0.0/

**Comment by llemeurfr** [comment]
I think it is time to stop using a `develop` branch on this project, which has no more planned releases and can be considered stable. The last fixes have been done on an extended period and needed to be merged in production asap. It was a burden to create a tag for each fix, but I agree we should do that using patch numbering.
In cas we need to add features to the project, it can safely be done using feature branches. 

I'll now check what could be missing from `develop` in `master`. When we're certain that we do not loose fixes from `develop`, I propose that we delete this `develop` branch. 

Re. squashing, is there a way using the Github UI to use it during PR merge?

**Comment by llemeurfr** [comment]
Now, an issue with deleting the `develop` branch is that currently, https://front-test.edrlab.org is associated with this branch (and https://front-prod.edrlab.org is associated with master). We cannot delete develop before this is modified.

**Comment by mickael-menu** [comment]
Thanks for the food for thought Daniel. I think we can all agree that unmanaged history is too messy to be useful.

I used to do `git pull --rebase origin develop` manually to keep a straight history while grouping PRs' commits (which is not possible with squash/rebase GitHub options). The rationale was that it's easier to find regressions using `git bisect` in smaller commits. But you got me thinking that it's probably best to have one self-contained commit for each PR, which has less chances of breaking stuff.

Yesterday I experimented with "Squash and merge" for a few PRs, and the results are quite nice:

<img width="1124" alt="Screenshot 2021-02-10 at 16 04 08" src="https://user-images.githubusercontent.com/58686775/107527954-9e117c80-6bb9-11eb-91a2-34b9426b8dd2.png">

This has also the added advantage that we control the final commit message even from external contributors.

Now for bigger PRs, maybe it would make more sense to squash internally to a few relevant commits before using "Rebase and merge" instead of squashing.

Anyways, I went ahead and disabled "Create a merge commit" from all the mobile repos 
<img width="838" alt="Screenshot 2021-02-10 at 16 05 17" src="https://user-images.githubusercontent.com/58686775/107528132-cac59400-6bb9-11eb-83c7-b83f62daaf09.png">

(@llemeurfr This answers your question)

> Re. squashing, is there a way using the Github UI to use it during PR merge?


Regarding https://www.conventionalcommits.org/en/v1.0.0/, that's an interesting concept and I see how it could be useful. However I find the generated changelogs not very user-friendly and a bit overwhelming. I think curating the changelogs is a must to keep them more useful than just `git log`.
